### PR TITLE
Fix Chunk Events not working correctly

### PR DESCRIPTION
**Pull request by nikita488**
It seems when porting patches from 1.12 to 1.13 the Chunk Events was messed up because of Chunk System rewrite.
## Problems:
**ChunkWatch.Watch** and **UnWatch** events was called for every chunk in player view but only once when player join to world while in 1.12 it fired when player actually start watching new chunk.
**ChunkDataEvent.Load** was not called at all.
**ChunkEvent.Load** was called every tick on client
**ChunkEvent.Unload** wasn't called for client at all.

This PR should fix this issues.

### CODE REVIEW

**Comment by nikita488** [comment]
Fixed merge conflict.

File: src/main/java/net/minecraftforge/event/ForgeEventFactory.java [hunk 1]
```diff
@@ -682,14 +682,6 @@ public static boolean saplingGrowTree(IWorld world, Random rand, BlockPos pos)
         return event.getResult() != Result.DENY;
     }
 
-    public static void fireChunkWatch(boolean watch, ServerPlayerEntity entity, ChunkPos chunkpos, ServerWorld world)
```
Comment (ichttt) [inline src/main/java/net/minecraftforge/event/ForgeEventFactory.java:685]: This is a breaking change as the method is public. Keep the method, but deprecate it and add a todo comment to remove the method in 1.15 if you no longer use it.

Comment (nikita488) [reply src/main/java/net/minecraftforge/event/ForgeEventFactory.java:685]: Fixed.

File: patches/minecraft/net/minecraft/client/multiplayer/ClientChunkProvider.java.patch [hunk 2]
```diff
@@ -1,18 +1,10 @@
 --- a/net/minecraft/client/multiplayer/ClientChunkProvider.java
 +++ b/net/minecraft/client/multiplayer/ClientChunkProvider.java
-@@ -61,6 +61,7 @@
-          int i = this.field_217256_d.func_217191_a(p_73234_1_, p_73234_2_);
-          Chunk chunk = this.field_217256_d.func_217192_a(i);
-          if (func_217249_a(chunk, p_73234_1_, p_73234_2_)) {
-+            net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.ChunkEvent.Unload(chunk));
-             this.field_217256_d.func_217190_a(i, chunk, (Chunk)null);
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/client/multiplayer/ClientChunkProvider.java.patch:8]: Moving this to where you did is wrong, as this event should be fired BEFORE the chunk is removed from the world.

File: patches/minecraft/net/minecraft/client/multiplayer/ClientChunkProvider.java.patch [hunk 3]
```diff
@@ -1,18 +1,10 @@
...
+             chunk.func_217326_a(p_217250_4_, p_217250_5_, p_217250_6_, p_217250_7_);
           }
- 
-@@ -113,6 +114,7 @@
-             worldlightmanager.func_215566_a(SectionPos.func_218154_a(p_217250_2_, j, p_217250_3_), ChunkSection.func_222628_a(chunksection));
-          }
- 
-+         net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.ChunkEvent.Load(chunk));
-          return chunk;
-       }
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/client/multiplayer/ClientChunkProvider.java.patch:17]: This move is also useless, as it is the same code flow, but you moved it above the chunk being finalized/lit for some unknown reason.

File: patches/minecraft/net/minecraft/world/server/ChunkManager.java.patch [hunk 4]
```diff
@@ -8,43 +8,43 @@
        }
  
        iprofiler.func_76319_b();
-@@ -398,6 +399,7 @@
-             if (this.field_219253_g.remove(p_219212_1_, p_219212_3_) && p_219185_5_ != null) {
-                if (p_219185_5_ instanceof Chunk) {
-                   ((Chunk)p_219185_5_).func_177417_c(false);
-+                  net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.ChunkEvent.Unload((Chunk)p_219185_5_));
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/world/server/ChunkManager.java.patch:15]: Again, you move this in a completely useless and breaking way. No.

File: patches/minecraft/net/minecraft/world/server/ChunkManager.java.patch [hunk 5]
```diff
@@ -8,43 +8,43 @@
...
- 
-                this.func_219229_a(p_219185_5_);
-@@ -475,6 +477,7 @@
+@@ -475,6 +476,7 @@
                 if (flag) {
                    IChunk ichunk = ChunkSerializer.func_222656_a(this.field_219255_i, this.field_219269_w, this.field_219260_n, p_223172_1_, compoundnbt);
                    ichunk.func_177432_b(this.field_219255_i.func_82737_E());
-+                  net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.ChunkEvent.Load(ichunk));
++                  net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.ChunkDataEvent.Load(ichunk, compoundnbt));
                    return Either.left(ichunk);
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/world/server/ChunkManager.java.patch:16]: This one actually looks correct.

File: patches/minecraft/net/minecraft/world/server/ChunkManager.java.patch [hunk 6]
```diff
@@ -8,43 +8,43 @@
...
-       for(int l = i - this.field_219247_A; l <= i + this.field_219247_A; ++l) {
-          for(int k = j - this.field_219247_A; k <= j + this.field_219247_A; ++k) {
-             ChunkPos chunkpos = new ChunkPos(l, k);
-+            net.minecraftforge.event.ForgeEventFactory.fireChunkWatch(p_219234_2_, p_219234_1_, chunkpos, field_219255_i);
-             this.func_219199_a(p_219234_1_, chunkpos, new IPacket[2], !p_219234_2_, p_219234_2_);
+@@ -714,6 +718,7 @@
+ 
+          if (!p_219199_5_ && p_219199_4_) {
+             p_219199_1_.func_213845_a(p_219199_2_);
++            net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.ChunkWatchEvent.UnWatch(p_219199_1_, p_219199_2_, field_219255_i));
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/world/server/ChunkManager.java.patch:39]: This looks correct, but you should probably move the Watch event here as well.

File: patches/minecraft/net/minecraft/world/server/ChunkManager.java.patch [hunk 7]
```diff
@@ -8,43 +8,43 @@
...
+          }
+ 
+       }
+@@ -1031,6 +1036,7 @@
           }
        }
+ 
++      net.minecraftforge.common.MinecraftForge.EVENT_BUS.post(new net.minecraftforge.event.world.ChunkWatchEvent.Watch(p_219180_1_, p_219180_3_.func_76632_l(), field_219255_i));
```
Comment (LexManos) [inline patches/minecraft/net/minecraft/world/server/ChunkManager.java.patch:47]: This is the wrong place, this has threading issues to consider. It should probably be moved into setChunkLoadedAtClient

**Review by LexManos — CHANGES_REQUESTED** [review]

**Comment by malte0811** [comment]
The watch/unwatch issue is causing trouble in the 1.14 port of IE. This PR seems to be abandoned, so I'm not sure what I'm supposed to do about it. If a second PR for just that issue is acceptable I'll make one.

**Pull request closed without merging**